Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp

```diff
@@ -245,34 +588,45 @@ void Client::CommunicateJobAvailability()
     m_StatusUpdateTimer.Start(); // reset time
 
     // has status changed since we last sent it?
-    uint32_t numJobsAvailable = (uint32_t)JobQueue::Get().GetNumDistributableJobsAvailable();
-    Protocol::MsgStatus msg( numJobsAvailable );
-
     MutexHolder mh( m_ServerListMutex );
     if ( m_ServerList.IsEmpty() )
     {
         return; // no servers to communicate with
     }
 
     // update each server to know how many jobs we have now
-    ServerState * it = m_ServerList.Begin();
-    const ServerState * const end = m_ServerList.End();
-    while ( it != end )
+    for ( size_t i=0; i<m_ServerList.GetSize(); ++i )
     {
-        if ( it->m_Connection )
+        ServerState * const ss = m_ServerList[ i ];
+        ASSERT( ss );
+
+        MutexHolder ssMH( ss->m_Mutex );
+        if ( ss->m_Connection )
         {
-            MutexHolder ssMH( it->m_Mutex );
-            if ( it->m_Connection )
+            uint32_t numJobsAvailable = 0;
+            uint32_t numJobsAvailableForWorker = 0;
+            JobQueue::Get().GetNumDistributableJobsAvailable(
+                    numJobsAvailable,
+                    numJobsAvailableForWorker );
+            if ( ss->m_NumJobsAvailableForWorker != numJobsAvailableForWorker )
             {
-                if ( it->m_NumJobsAvailable != numJobsAvailable )
-                {
-                    PROFILE_SECTION( "UpdateJobAvailability" )
-                    SendMessageInternal( it->m_Connection, msg );
-                    it->m_NumJobsAvailable = numJobsAvailable;
-                }
+                PROFILE_SECTION( "UpdateJobAvailability" )
+                Protocol::MsgStatus msg( numJobsAvailableForWorker );
+                SendMessageInternal( ss->m_Connection, msg );
+                ss->m_NumJobsAvailableForWorker = numJobsAvailableForWorker;
+            }
+            if ( numJobsAvailable > 0 &&  // if any job in queue
+                 numJobsAvailableForWorker == 0 )  // worker cannot build any job
```

Comment:
numJobsAvailableForWorker can only ever be less than numJobsAvailable, so it seems we only need to check one